### PR TITLE
Libcbor integrate

### PR DESCRIPTION
**Pull request by TingDaoK**
*Issue #, if available:*

*Description of changes:*

- Integrate with libcbor to provide a cbor encode/decode interface based on aws-c-common. Major decision made:
  - Decoding will have peek_type/get_next_\*, and peek type will cache the next element, while get_next_\* will consume the cached one, or just decode the next one if no cache was stored. 
  - Separate type for indefinite map/array for easier handling
  - Separate type for indefinite string/bytes to avoid additional copy (The chunks has to be copied if we want to concatenate them)

- Debatable: have the cbor in aws-c-common or aws-c-sdkutils, I put it in aws-c-common mostly for:
  - common holds all the external 3rd-party lib for now
  - right now, cbor only required by SDKs, but it's still a generic concept not limited to aws.

- TODO:
  - ~Conversion from double to float is tricky, numbers like "1.1" cannot be represented by double or float.~
  - ~Tags, should we be helpful to convert the tag value of timestamp (probably big number in the future) for the user? Or, let user to read the tag value and decide what to do with it??~
  - ~Error handling. Error code, logging, documentation, etc.~
  - ~Fuzz test from our side as well to make sure not undefined behavior~
  - Consider:
    - Add separate encoding for float
    - Add decode number to float, decode number to double
    - Add helper to handle indefinite string/bytes

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/aws/common/byte_buf.h [hunk 1]
```diff
@@ -451,6 +451,15 @@ int aws_byte_buf_append_and_update(struct aws_byte_buf *to, struct aws_byte_curs
...
 int aws_byte_buf_append_null_terminator(struct aws_byte_buf *buf);
 
+/**
+ * Expand the buffer appropriately to meet the requested capacity.
+ *
+ * If the the buffer's capacity is currently larger than the request capacity, the
+ * function does nothing (no shrink is performed).
+ */
+AWS_COMMON_API
+int aws_byte_buf_reserve_dynamic(struct aws_byte_buf *buffer, size_t requested_capacity);
```
Comment (DmitriyMusatkin) [inline include/aws/common/byte_buf.h:461]: i actually run into same problem couple days ago and could have sworn we already had helper. So was thinking of implementing it. 

Minor suggestion on naming: maybe something like aws_byte_buf_ensure_capacity?

File: source/byte_buf.c [hunk 2]
```diff
@@ -778,31 +778,53 @@ int aws_byte_buf_append_byte_dynamic_secure(struct aws_byte_buf *buffer, uint8_t
...
 
     if (requested_capacity <= buffer->capacity) {
         AWS_POSTCONDITION(aws_byte_buf_is_valid(buffer));
         return AWS_OP_SUCCESS;
     }
-    if (!buffer->buffer && !buffer->capacity && requested_capacity > buffer->capacity) {
-        if (aws_byte_buf_init(buffer, buffer->allocator, requested_capacity)) {
+
+    size_t new_capacity = requested_capacity;
+    if (dynamic_expand) {
```
[the remark this answers is not in the document]
Comment (TingDaoK) [reply source/byte_buf.c:791]: then it will just expand it to the requested_capacity. I basically follow the logic from append_dynamic.

File: include/aws/common/cbor.h [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+
+/**
+ * @brief Consume the next data item, includes all the content within the data item.
+ * Notes: this function will not ensure the data item is well-formed.
+ *
+ * @param src The src to parse data from
+ * @return AWS_OP_SUCCESS successfully consumed the next data item, otherwise AWS_OP_ERR.
+ */
+AWS_COMMON_API
+int aws_cbor_decode_consume_next_data_item(struct aws_cbor_decoder *decoder);
```
[the remark this answers is not in the document]
Comment (TingDaoK) [reply include/aws/common/cbor.h:332]: yeah, I'll add example to the documentation.

File: include/aws/common/macros.h [hunk 4]
```diff
@@ -131,6 +131,24 @@ enum { AWS_CACHE_LINE = 64 };
 #    define AWS_SUPPRESS_TSAN
 #endif
 
+#if defined(__has_feature)
```
[the remark this answers is not in the document]
Comment (TingDaoK) [reply include/aws/common/macros.h:134]: hmmm, I am thinking since we already have `AWS_SUPPRESS_TSAN` and `AWS_SUPPRESS_ASAN`, probably makes sense to have this as well. but, yeah, we are not using it for now.

File: include/aws/common/cbor.h [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+ *  - result of value not supported.
+ */
+enum aws_cbor_element_type {
+    AWS_CBOR_TYPE_UINT,
+    AWS_CBOR_TYPE_NEGINT,
+    AWS_CBOR_TYPE_DOUBLE,
+    AWS_CBOR_TYPE_BYTESTRING,
+    AWS_CBOR_TYPE_STRING,
+
+    AWS_CBOR_TYPE_ARRAY_START,
```
[the remark this answers is not in the document]
Comment (TingDaoK) [reply include/aws/common/cbor.h:40]: I think it may make more sense to indicate the element only tells the array starts, not the content of the array.

File: include/aws/common/cbor.h [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+#include <aws/common/byte_buf.h>
+#include <aws/common/common.h>
+
+AWS_PUSH_SANE_WARNING_LEVEL
+AWS_EXTERN_C_BEGIN
+
+/**
+ * The element types use by peek type.
+ * An extension for cbor major type in RFC8949 section 3.1
+ * Major type 0 - AWS_CBOR_TYPE_UINT
```
[the remark this answers is not in the document]
Comment (TingDaoK) [reply include/aws/common/cbor.h:18]: the decoding API for get_next_uint_val will read the header and the following bytes if any for the unsigned value it indicates (includes less than 24, and short, int and long) to and uint64_t. Yeah, we could add to it if we want to expose more specific encode/decode, maybe another "peek_width/with_width" API on top.

File: include/aws/common/cbor.h [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+ *
+ * @param allocator
+ * @param init_buffer_size The initial size of the temporary buffer, which will grow as needed.
+ * @return aws_cbor_encoder
+ */
+AWS_COMMON_API
+struct aws_cbor_encoder *aws_cbor_encoder_new(struct aws_allocator *allocator, size_t init_buffer_size);
+
+AWS_COMMON_API
+struct aws_cbor_encoder *aws_cbor_encoder_release(struct aws_cbor_encoder *encoder);
```
[the remark this answers is not in the document]
Comment (TingDaoK) [reply include/aws/common/cbor.h:99]: we usually use destroy, but I am still not super sure will we have refcount for the encoder/decoder or not.

File: source/byte_buf.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (DmitriyMusatkin) [reply source/byte_buf.c:791]: can you add comment that dynamic expand does max(2*current_capacity,  requested_capacity) and non-dynamic just does requested capacity. 

might be worth adding to header is that dynamic expand is preferred for when we know we are going to add a lot of tiny elements.

[the remark this answers is not in the document]
Comment (DmitriyMusatkin) [reply source/byte_buf.c:791]: goes back to my comment in the header, that reserve dynamic is a bit of confusing naming. my initial reading was that it referred to some difference between static and dynamic buffers

File: tests/cbor_test.c [hunk 8]
```diff
@@ -0,0 +1,532 @@
+/**
```
Comment (DmitriyMusatkin) [inline tests/cbor_test.c:1]: need test for new buffer functionality as well

File: tests/cbor_test.c [hunk 9]
```diff
@@ -0,0 +1,532 @@
...
+#include <float.h>
+#include <math.h>
+
+#include <aws/testing/aws_test_harness.h>
+
+#define CBOR_TEST_CASE(NAME)                                                                                           \
+    AWS_TEST_CASE(NAME, s_test_##NAME);                                                                                \
+    static int s_test_##NAME(struct aws_allocator *allocator, void *ctx)
+
+CBOR_TEST_CASE(cbor_encode_decode_int_test) {
```
Comment (DmitriyMusatkin) [inline tests/cbor_test.c:16]: you want to test that uint64 max value is written out correctly as well.

File: tests/cbor_test.c [hunk 10]
```diff
@@ -0,0 +1,532 @@
...
+     * uint16_t max to uint32_t maxx takes 5 bytes
+     * uint32_t max to uint64_t max takes 9 bytes
+     */
+    uint64_t values[VALUE_NUM] = {23, 24, UINT8_MAX + 1, UINT16_MAX + 1U, UINT32_MAX + 1LLU};
+    uint64_t expected_encoded_len[VALUE_NUM] = {1, 2, 3, 5, 9};
+
+    size_t encoded_len = 0;
+    struct aws_cbor_encoder *encoder = aws_cbor_encoder_new(allocator, 128);
+    /* Unsigned int */
+    for (size_t i = 0; i < VALUE_NUM; i++) {
```
Comment (DmitriyMusatkin) [inline tests/cbor_test.c:35]: nit: we have ARRAYSIZE_OF macro somewhere that you can use to avoid having a constant

File: source/cbor.c [hunk 11]
```diff
@@ -0,0 +1,708 @@
...
+
+struct aws_cbor_encoder {
+    struct aws_allocator *allocator;
+
+    /* Refcount maybe */
+    struct aws_ref_count refcount;
+    struct aws_byte_buf encoded_buf;
+};
+
+struct aws_cbor_encoder *aws_cbor_encoder_new(struct aws_allocator *allocator, size_t init_buffer_size) {
```
Comment (DmitriyMusatkin) [inline source/cbor.c:42]: would users know whats a good size they can pass in for init buffer size. maybe we just pick default for them

Comment (DmitriyMusatkin) [reply source/cbor.c:42]: another thing to consider is should we allow them to pass a buffer they have already allocated?

File: tests/fuzz/cbor_double_encode_decode.c [hunk 12]
```diff
@@ -0,0 +1,66 @@
+/**
```
Comment (DmitriyMusatkin) [inline tests/fuzz/cbor_double_encode_decode.c:1]: does llvm fuzzer generate completely random data here as input? or it has some idea of what values are valid in cbor?

File: source/cbor.c [hunk 13]
```diff
@@ -0,0 +1,708 @@
...
+}
+
+void aws_cbor_encode_double(struct aws_cbor_encoder *encoder, double value) {
+    if (isnan(value) || isinf(value)) {
+        /* For special value: NAN/INFINITY, type cast to float and encode into single float. */
+        aws_cbor_encode_single_float(encoder, (float)value);
+        return;
+    }
+    /* Conversation from int to floating-type is implementation defined if loss of precision */
+    if (value <= (double)INT64_MAX && value >= (double)INT64_MIN) {
```
[the remark this answers is not in the document]
Comment (TingDaoK) [reply source/cbor.c:115]: Just want to simplify the implementation. 

I guess in real life if we cast double to 64bit int, we don't actually have benefit to save any bytes here.

File: tests/cbor_test.c (hunk 10 above)
Comment (TingDaoK) [reply tests/cbor_test.c:35]: yeah, arguable which is better as here, we have multiple array that I want to make sure they all the same size

File: tests/fuzz/cbor_double_encode_decode.c (hunk 12 above)
Comment (TingDaoK) [reply tests/fuzz/cbor_double_encode_decode.c:1]: I believe it is completely random, https://llvm.org/docs/LibFuzzer.html#id23

But, for sure, it doesn't know anything about cbor

File: source/cbor.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (DmitriyMusatkin) [reply source/cbor.c:115]: im fine on losing optimization opportunity here since its a corner case that is not likely to occur in practice, but we should doc that it was done in order to simplify code.

File: tests/fuzz/cbor_double_encode_decode.c (hunk 12 above)
Comment (DmitriyMusatkin) [reply tests/fuzz/cbor_double_encode_decode.c:1]: so it will ignore anything malformed? how often will it generate malformed garbage vs something thats parsable?

just want to verify we are actually getting useful results out of it

File: include/aws/common/cbor.h [hunk 14]
```diff
@@ -0,0 +1,450 @@
...
+ *  - 25/26/27 - AWS_CBOR_TYPE_DOUBLE
+ *  - 31 - AWS_CBOR_TYPE_BREAK
+ *  - result of value not supported.
+ */
+enum aws_cbor_element_type {
+    AWS_CBOR_TYPE_UINT,
+    AWS_CBOR_TYPE_NEGINT,
+    AWS_CBOR_TYPE_DOUBLE,
+    AWS_CBOR_TYPE_BYTESTRING,
+    AWS_CBOR_TYPE_STRING,
```
Comment (graebm) [inline include/aws/common/cbor.h:38]: trivial: weird to do BYTESTRING but then just STRING

maybe: BYTESTRING / TEXTSTRING to stick with the official terminology

or: BYTES / TEXT to keep it short

File: include/aws/common/cbor.h [hunk 15]
```diff
@@ -0,0 +1,450 @@
...
+/**
+ * @brief Encode a AWS_CBOR_TYPE_BYTESTRING value to "smallest possible" in encoder's buffer.
+ *  Referring to RFC8949 section 4.2.1, the length of "from" will be encoded first and then the value of "from" will
+ * be followed.
+ *
+ * @param encoder
+ * @param from value to encode.
+ */
+AWS_COMMON_API
+void aws_cbor_encoder_write_bytes(struct aws_cbor_encoder *encoder, const struct aws_byte_cursor from);
```
Comment (graebm) [inline include/aws/common/cbor.h:165]: trivial weird that the enum is BYTESTRING but this function is write_bytes(). Keep it consistent

BYTESTRING / write_bytestring()
or
BYTES / write_bytes()

File: include/aws/common/cbor.h [hunk 16]
```diff
@@ -0,0 +1,450 @@
...
+AWS_COMMON_API
+void aws_cbor_encoder_write_tag(struct aws_cbor_encoder *encoder, uint64_t tag_number);
+
+/**
+ * @brief Encode a simple value AWS_CBOR_TYPE_NULL
+ *
+ * @param encoder
+ */
+AWS_COMMON_API
+void aws_cbor_encoder_write_null(struct aws_cbor_encoder *encoder);
```
Comment (graebm) [inline include/aws/common/cbor.h:222]: trivial/debatable: worth naming like `write_simple_null()` to group all the simples? Or just have like `write_simple(struct aws_cbor_encoder *encoder, uint64_t simple_value)`? (I'm not sure if user-defined simple values are a thing...)

I guess that would allow users to pass illegal simple values, but this API also allows users to encode their arrays and maps wrong too, so we could just naively forge ahead and encode whatever they say and that's at least consistent API-wise

File: include/aws/common/cbor.h [hunk 17]
```diff
@@ -0,0 +1,450 @@
...
+ * All other type will result in error.
+ *
+ * Notes: It's user's responsibility to track and add the break to
+ * close the corresponding inf_start
+ *
+ * @param encoder
+ * @param type
+ */
+AWS_COMMON_API
+int aws_cbor_encoder_write_inf_start(struct aws_cbor_encoder *encoder, enum aws_cbor_element_type type);
```
Comment (graebm) [inline include/aws/common/cbor.h:264]: This is the only write() function that isn't `void`

I'd suggest keeping it simple/consistent by just having 4 separate functions for bytes/text/array/map

Or just naively encode the bad type and continue on our merry way, since there are plenty of other ways for users to encode data illegally

**Review by graebm — COMMENTED** [review]
I only looked at the encoder part of the public header. Apologies for naive comments, since I haven't looked at the decoder yet, or any implementation details...

File: include/aws/common/cbor.h [hunk 18]
```diff
@@ -0,0 +1,450 @@
...
+    AWS_CBOR_TYPE_MAP_START,
+
+    AWS_CBOR_TYPE_TAG,
+
+    AWS_CBOR_TYPE_BOOL,
+    AWS_CBOR_TYPE_NULL,
+    AWS_CBOR_TYPE_UNDEFINE,
+    AWS_CBOR_TYPE_BREAK,
+
+    AWS_CBOR_TYPE_INF_BYTESTRING_START,
```
Comment (graebm) [inline include/aws/common/cbor.h:50]: debatable: it's "indefinite", not "infinite". Maybe abbreviate as INDEF?

File: include/aws/common/cbor.h [hunk 19]
```diff
@@ -0,0 +1,450 @@
...
+AWS_COMMON_API
+int aws_cbor_decoder_pop_next_neg_val(struct aws_cbor_decoder *decoder, uint64_t *out);
+AWS_COMMON_API
+int aws_cbor_decoder_pop_next_double_val(struct aws_cbor_decoder *decoder, double *out);
+AWS_COMMON_API
+int aws_cbor_decoder_pop_next_boolean_val(struct aws_cbor_decoder *decoder, bool *out);
+AWS_COMMON_API
+int aws_cbor_decoder_pop_next_bytes_val(struct aws_cbor_decoder *decoder, struct aws_byte_cursor *out);
+AWS_COMMON_API
+int aws_cbor_decoder_pop_next_str_val(struct aws_cbor_decoder *decoder, struct aws_byte_cursor *out);
```
[the remark this answers is not in the document]
Comment (graebm) [reply include/aws/common/cbor.h:398]: maybe "text" since CBOR calls them "byte strings" and "text strings"

File: include/aws/common/cbor.h [hunk 20]
```diff
@@ -0,0 +1,450 @@
...
+ *      416d74 --   "Amt"
+ *   21        -- Second value, -2
+ *   FF        -- "break"
+ *
+ * @param decoder The decoder to parse data from
+ * @param consumed_type Optional, get the type of the consumed element.
+ * @return AWS_OP_SUCCESS successfully consumed the next element, otherwise AWS_OP_ERR.
+ */
+AWS_COMMON_API
+int aws_cbor_decoder_consume_next_element(struct aws_cbor_decoder *decoder, enum aws_cbor_element_type *consumed_type);
```
[the remark this answers is not in the document]
Comment (TingDaoK) [reply include/aws/common/cbor.h:371]: will remove the consumed_type

File: source/cbor.c [hunk 21]
```diff
@@ -0,0 +1,739 @@
...
+    } cbor_data;
+};
+
+struct aws_cbor_decoder {
+    struct aws_allocator *allocator;
+
+    struct aws_byte_cursor src;
+    /* Keep the original src to revert decoded part. */
+    struct aws_byte_cursor original_src;
+    uint64_t bytes_decoded;
```
[the remark this answers is not in the document]
Comment (graebm) [reply source/cbor.c:285]: actually, I don't see this value used ANYWHERE, just gets incremented in one place. delete?
```suggestion
```

File: source/cbor.c [hunk 22]
```diff
@@ -0,0 +1,739 @@
...
+overflow:
+    /* Overflow detected, it's technically not a malformed cbor item, but for this helper */
+    AWS_LOGF_ERROR(AWS_LS_COMMON_CBOR, "The decoded timestamp cannot fit into int64_t.");
+    /* TODO: do we want to revert the decoded part???? Probably??? */
+    return aws_raise_error(AWS_ERROR_OVERFLOW_DETECTED);
+}
+
+int aws_cbor_decoder_consume_next_data_item(struct aws_cbor_decoder *decoder) {
+
+    if (decoder->cached_context.type == AWS_CBOR_TYPE_MAX) {
```
Comment (graebm) [inline source/cbor.c:658]: should we be checking `decoder->error_code` first thing? or is that always handled first-thing by the helper functions?

File: source/cbor.c [hunk 23]
```diff
@@ -0,0 +1,739 @@
...
+        }                                                                                                              \
+        if (s_cbor_decode_next_element(decoder)) {                                                                     \
+            return AWS_OP_ERR;                                                                                         \
+        }                                                                                                              \
+    decode_done:                                                                                                       \
+        if ((decoder)->cached_context.type != (expected_cbor_type)) {                                                  \
+            AWS_LOGF_ERROR(                                                                                            \
+                AWS_LS_COMMON_CBOR,                                                                                    \
+                "The decoder got unexpected type: %d, while expecting type: %d.",                                      \
```
Comment (graebm) [inline source/cbor.c:509]: might be worth a `const char *aws_cbor_type_cstr(aws_cbor_type type)` helper function, that you can use here, so the error message is more understandable to otehrs. Similar to https://github.com/awslabs/aws-c-io/blob/375f393ba6b900e7c3a46f9f5a0eebf2f7a22073/source/pkcs11_lib.c#L301-L309

"The decoder got unexpected type: 0, while expecting type: 2."
vs
"The decoder got unexpected type: 0 (AWS_CBOR_TYPE_UINT), while expecting type: 2 (AWS_CBOR_TYPE_FLOAT)."

File: include/aws/common/cbor.h [hunk 24]
```diff
@@ -0,0 +1,457 @@
...
+ * (Order with priority) when the conversation will not cause precision loss.
+ *
+ * @param encoder
+ * @param value value to encode.
+ */
+AWS_COMMON_API
+void aws_cbor_encoder_write_float(struct aws_cbor_encoder *encoder, double value);
+
+/**
+ * @brief Encode a AWS_CBOR_TYPE_BYTES value to "smallest possible" in encoder's buffer.
```
[the remark this answers is not in the document]
Comment (TingDaoK) [reply include/aws/common/cbor.h:149]: bytes still have a integer for the number of bytes, which still follows the "smallest possible" rule.

File: include/aws/common/cbor.h [hunk 25]
```diff
@@ -0,0 +1,450 @@
...
+ * @brief Get the next AWS_CBOR_TYPE_ARRAY_START element. Only consume the AWS_CBOR_TYPE_ARRAY_START element and set the
+ * size of array to *out_size, not the content of the array. The next *out_size cbor data items will be the content of
+ * the array for a valid cbor data,
+ *
+ * @param decoder
+ * @param out_size store the size of array if succeed.
+ * @return AWS_OP_SUCCESS successfully consumed the next element and get the result, otherwise AWS_OP_ERR.
+ */
+AWS_COMMON_API
+int aws_cbor_decoder_pop_next_array_start(struct aws_cbor_decoder *decoder, uint64_t *out_size);
```
[the remark this answers is not in the document]
Comment (TingDaoK) [reply include/aws/common/cbor.h:417]: thought we talked about it.
```
    AWS_CBOR_TYPE_INDEF_BYTES_START,
    AWS_CBOR_TYPE_INDEF_TEXT_START,
    AWS_CBOR_TYPE_INDEF_ARRAY_START,
    AWS_CBOR_TYPE_INDEF_MAP_START,
```

We have separate types for the indef types. I can add notes to the documents.

File: source/byte_buf.c [hunk 26]
```diff
@@ -778,31 +778,53 @@ int aws_byte_buf_append_byte_dynamic_secure(struct aws_byte_buf *buffer, uint8_t
...
-    buffer->capacity = requested_capacity;
+    buffer->capacity = new_capacity;
 
     AWS_POSTCONDITION(aws_byte_buf_is_valid(buffer));
     return AWS_OP_SUCCESS;
 }
 
+int aws_byte_buf_ensure_capacity(struct aws_byte_buf *buffer, size_t requested_capacity) {
+    return s_aws_byte_buf_reserve(buffer, requested_capacity, true /*dynamic_expand*/);
+}
```
[the remark this answers is not in the document]
Comment (TingDaoK) [reply source/byte_buf.c:822]: https://github.com/awslabs/aws-c-common/pull/1101#discussion_r1561753179

File: source/cbor.c [hunk 27]
```diff
@@ -0,0 +1,739 @@
...
+    if (decoder->cached_context.type == AWS_CBOR_TYPE_MAX) {
+        /* There was no cache, decode the next item */
+        if (s_cbor_decode_next_element(decoder)) {
+            return AWS_OP_ERR;
+        }
+    }
+    switch (decoder->cached_context.type) {
+        case AWS_CBOR_TYPE_TAG:
+            /* Read the next data item */
+            /* TODO: error check for the tag content?? */
```
Comment (graebm) [inline source/cbor.c:667]: Should we do this before commit, or will there be a follow-up?

File: include/aws/common/cbor.h [hunk 28]
```diff
@@ -0,0 +1,457 @@
...
+    AWS_CBOR_TYPE_NULL,
+    AWS_CBOR_TYPE_UNDEFINED,
+    AWS_CBOR_TYPE_BREAK,
+
+    AWS_CBOR_TYPE_INDEF_BYTES_START,
+    AWS_CBOR_TYPE_INDEF_TEXT_START,
+    AWS_CBOR_TYPE_INDEF_ARRAY_START,
+    AWS_CBOR_TYPE_INDEF_MAP_START,
+
+    AWS_CBOR_TYPE_MAX,
```
[the remark this answers is not in the document]
Comment (graebm) [reply include/aws/common/cbor.h:55]: Or `_UNKNOWN = 0` so that a zeroed out struct gets a good sensible default

File: source/cbor.c [hunk 29]
```diff
@@ -0,0 +1,739 @@
...
+            decoder->cached_context.type = AWS_CBOR_TYPE_MAX;
+            if (aws_cbor_decoder_consume_next_data_item(decoder)) {
+                return AWS_OP_ERR;
+            }
+            break;
+        case AWS_CBOR_TYPE_MAP_START: {
+            uint64_t num_map_item = decoder->cached_context.cbor_data.map_start;
+            /* Reset type */
+            decoder->cached_context.type = AWS_CBOR_TYPE_MAX;
+            for (uint64_t i = 0; i < num_map_item; i++) {
```
Comment (graebm) [inline source/cbor.c:677]: Could this be a vector for some kind of attack? Like, attacker sends CBOR with a map or array that says it's UINT64_MAX items. Do we pre-allocate space anywhere, that could lead to a out-of-memory crash? Or do we loop through all items?

Or it's not really possible to attack that way on decode? Because we just decode data as it comes in, and they'd actually have to send all that data (not just claim they were GOING to send all that data) ???

File: include/aws/common/cbor.h [hunk 30]
```diff
@@ -0,0 +1,450 @@
...
+ * Major type 6 - AWS_CBOR_TYPE_TAG
+ * Major type 7
+ *  - 20/21 - AWS_CBOR_TYPE_BOOL
+ *  - 22 - AWS_CBOR_TYPE_NULL
+ *  - 23 - AWS_CBOR_TYPE_UNDEFINE
+ *  - 25/26/27 - AWS_CBOR_TYPE_DOUBLE
+ *  - 31 - AWS_CBOR_TYPE_BREAK
+ *  - result of value not supported.
+ */
+enum aws_cbor_element_type {
```
[the remark this answers is not in the document]
Comment (graebm) [reply include/aws/common/cbor.h:33]: aws_cbor_api_type?

[the remark this answers is not in the document]
Comment (graebm) [reply include/aws/common/cbor.h:33]: aws_cbor_decode_type?

[the remark this answers is not in the document]
Comment (graebm) [reply include/aws/common/cbor.h:33]: or just aws_cbor_type and say "it's not 1:1 with major type"

File: source/cbor.c [hunk 31]
```diff
@@ -0,0 +1,739 @@
...
+        if (value == (double)int_value) {
+            if (int_value < 0) {
+                aws_cbor_encoder_write_negint(encoder, (uint64_t)(-1 - int_value));
+            } else {
+                aws_cbor_encoder_write_uint(encoder, (uint64_t)(int_value));
+            }
+            return;
+        }
+    }
+    if (value <= FLT_MAX && value >= -FLT_MAX) {
```
[the remark this answers is not in the document]
Comment (TingDaoK) [reply source/cbor.c:141]: It's because of undefined behavior. UBSAN will flag this when the conversion is outside the range. 

`FLT_MIN` is "Minimum representable **positive** floating-point number.". Only reason I put this check is to avoid UBSAN, I have tried [this](https://github.com/awslabs/aws-c-common/pull/1101/commits/f0b8a99ee0c4ead16df56c076d3314b464860048), but I am more happy about -max to max which is enough to silent the UBSAN.

File: include/aws/common/cbor.h (hunk 30 above)
[the remark this answers is not in the document]
Comment (TingDaoK) [reply include/aws/common/cbor.h]: Do we also want to change naming of `aws_cbor_decoder_consume_next_element`?

File: source/cbor.c (hunk 27 above)
Comment (TingDaoK) [reply source/cbor.c:691]: Oh, I have a notes [here](https://github.com/awslabs/aws-c-common/pull/1101/files#diff-fcdda971f4f90ee939cc56bc2f2f3faa5aabbd468f17958e1be8ab913287d707R354), maybe I'll just remove this TODO.

File: include/aws/common/cbor.h (hunk 30 above)
[the remark this answers is not in the document]
Comment (graebm) [reply include/aws/common/cbor.h]: `aws_cbor_decoder_consume_next_single() / aws_cbor_decoder_consume_next_recursive()`
`aws_cbor_decoder_consume_next_singlestep() / aws_cbor_decoder_consume_next_recursively()`
?

**Review by graebm — APPROVED** [review]
